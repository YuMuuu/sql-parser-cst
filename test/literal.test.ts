import { dialect, parseExpr, testExpr } from "./test_utils";

describe("literal", () => {
  it("single-quoted string", () => {
    expect(parseExpr(`'hello'`)).toMatchSnapshot();
  });
  it("single-quoted string with escapes", () => {
    expect(parseExpr(`'hel\\'lo'`)).toMatchSnapshot();
    expect(parseExpr(`'hel''lo'`)).toMatchSnapshot();
  });
  it("single-quoted string with charset", () => {
    expect(parseExpr(`_binary 'hello'`)).toMatchSnapshot();
  });
  it("single-quoted string with charset and comment", () => {
    expect(parseExpr(`_latin1 /* comment */ 'hello'`)).toMatchSnapshot();
  });

  dialect("mysql", () => {
    it("double-quoted string", () => {
      expect(parseExpr(`"hello"`)).toMatchInlineSnapshot(`
        {
          "text": ""hello"",
          "type": "string",
        }
      `);
    });
    it("double-quoted string with escapes", () => {
      testExpr(`"hel\\"lo"`);
      testExpr(`"hel""lo"`);
    });
    it("double-quoted string with charset", () => {
      expect(parseExpr(`_latin1"hello"`)).toMatchInlineSnapshot(`
        {
          "charset": "latin1",
          "string": {
            "text": ""hello"",
            "type": "string",
          },
          "type": "string_with_charset",
        }
      `);
    });
    it("double-quoted string with charset and comments", () => {
      testExpr(`_latin1 -- comment1\n -- comment2\n 'hello'`);
    });
  });

  it("hex literal", () => {
    expect(parseExpr(`0xAAFF11`)).toMatchSnapshot();
  });
  it("hex literal with charset", () => {
    expect(parseExpr(`_utf8 0xAAFF11`)).toMatchSnapshot();
  });

  it("bit string", () => {
    expect(parseExpr(`b'011001'`)).toMatchSnapshot();
  });
  it("bit string with charset", () => {
    expect(parseExpr(`_big5 b'011001'`)).toMatchSnapshot();
  });

  it("hex string", () => {
    expect(parseExpr(`x'AFC123'`)).toMatchSnapshot();
  });
  it("hex string with charset", () => {
    expect(parseExpr(`_utf16le X'AFC123'`)).toMatchSnapshot();
  });

  it("natural character set string", () => {
    expect(parseExpr(`N'hello'`)).toMatchSnapshot();
  });
  it("natural character set string with escapes", () => {
    expect(parseExpr(`n'hel\\'lo'`)).toMatchSnapshot();
    expect(parseExpr(`N'hel''lo'`)).toMatchSnapshot();
  });

  it("boolean", () => {
    expect(parseExpr(`true`)).toMatchSnapshot();
    expect(parseExpr(`TRUE`)).toMatchSnapshot();
    expect(parseExpr(`false`)).toMatchSnapshot();
    expect(parseExpr(`FALSE`)).toMatchSnapshot();
  });

  it("null", () => {
    expect(parseExpr(`null`)).toMatchSnapshot();
    expect(parseExpr(`NULL`)).toMatchSnapshot();
  });

  it("number", () => {
    expect(parseExpr(`0`)).toMatchSnapshot();
    expect(parseExpr(`123`)).toMatchSnapshot();
    expect(parseExpr(`0.15`)).toMatchSnapshot();
    expect(parseExpr(`123e15`)).toMatchSnapshot();
    expect(parseExpr(`1.23E+13`)).toMatchSnapshot();
    expect(parseExpr(`0.107e-62`)).toMatchSnapshot();
    expect(parseExpr(`-6`)).toMatchSnapshot();
    expect(parseExpr(`+18`)).toMatchSnapshot();
  });

  it("datetime", () => {
    testExpr(`TIME '20:15:00'`);
    testExpr(`DATE "1995-06-01"`);
    testExpr(`DATEtime '1995-06-01 20:15:00'`);
    testExpr(`timestamp "1995-06-01 20:15:00"`);
    testExpr(`DATETIME /* com1 */ '20:15:00'`);
  });

  dialect("mysql", () => {
    it("datetime with double-quoted string", () => {
      testExpr(`TIME "20:15:00"`);
      testExpr(`DATE "1995-06-01"`);
      testExpr(`DATEtime "1995-06-01 20:15:00"`);
      testExpr(`timestamp "1995-06-01 20:15:00"`);
    });
  });
});
