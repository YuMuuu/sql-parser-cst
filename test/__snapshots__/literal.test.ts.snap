// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`literal bit string 1`] = `
{
  "text": "b'011001'",
  "type": "string",
}
`;

exports[`literal bit string with charset 1`] = `
{
  "charset": "big5",
  "string": {
    "text": "b'011001'",
    "type": "string",
  },
  "type": "string_with_charset",
}
`;

exports[`literal boolean 1`] = `
{
  "text": "true",
  "type": "bool",
}
`;

exports[`literal boolean 2`] = `
{
  "text": "TRUE",
  "type": "bool",
}
`;

exports[`literal boolean 3`] = `
{
  "text": "false",
  "type": "bool",
}
`;

exports[`literal boolean 4`] = `
{
  "text": "FALSE",
  "type": "bool",
}
`;

exports[`literal datetime 1`] = `
{
  "kw": {
    "text": "TIME",
    "type": "keyword",
  },
  "string": {
    "text": "'20:15:00'",
    "type": "string",
  },
  "type": "datetime",
}
`;

exports[`literal datetime 2`] = `
{
  "kw": {
    "text": "DATE",
    "type": "keyword",
  },
  "string": {
    "text": ""1995-06-01"",
    "type": "string",
  },
  "type": "datetime",
}
`;

exports[`literal datetime 3`] = `
{
  "kw": {
    "text": "DATEtime",
    "type": "keyword",
  },
  "string": {
    "text": "'1995-06-01 20:15:00'",
    "type": "string",
  },
  "type": "datetime",
}
`;

exports[`literal datetime 4`] = `
{
  "kw": {
    "text": "timestamp",
    "type": "keyword",
  },
  "string": {
    "text": ""1995-06-01 20:15:00"",
    "type": "string",
  },
  "type": "datetime",
}
`;

exports[`literal datetime with comment 1`] = `
{
  "kw": {
    "text": "DATETIME",
    "type": "keyword",
  },
  "string": {
    "leadingComments": [
      {
        "text": "/* com1 */",
        "type": "block_comment",
      },
    ],
    "text": "'20:15:00'",
    "type": "string",
  },
  "type": "datetime",
}
`;

exports[`literal double-quoted string 1`] = `
{
  "text": ""hello"",
  "type": "string",
}
`;

exports[`literal double-quoted string with charset 1`] = `
{
  "charset": "latin1",
  "string": {
    "text": ""hello"",
    "type": "string",
  },
  "type": "string_with_charset",
}
`;

exports[`literal double-quoted string with charset and comments 1`] = `
{
  "charset": "latin1",
  "string": {
    "leadingComments": [
      {
        "text": "-- comment1",
        "type": "line_comment",
      },
      {
        "text": "-- comment2",
        "type": "line_comment",
      },
    ],
    "text": "'hello'",
    "type": "string",
  },
  "type": "string_with_charset",
}
`;

exports[`literal double-quoted string with escapes 1`] = `
{
  "text": ""hel\\"lo"",
  "type": "string",
}
`;

exports[`literal double-quoted string with escapes 2`] = `
{
  "text": ""hel"",
  "type": "string",
}
`;

exports[`literal hex literal 1`] = `
{
  "text": "0xAAFF11",
  "type": "string",
}
`;

exports[`literal hex literal with charset 1`] = `
{
  "charset": "utf8",
  "string": {
    "text": "0xAAFF11",
    "type": "string",
  },
  "type": "string_with_charset",
}
`;

exports[`literal hex string 1`] = `
{
  "text": "x'AFC123'",
  "type": "string",
}
`;

exports[`literal hex string with charset 1`] = `
{
  "charset": "utf16le",
  "string": {
    "text": "X'AFC123'",
    "type": "string",
  },
  "type": "string_with_charset",
}
`;

exports[`literal natural character set string 1`] = `
{
  "text": "N'hello'",
  "type": "string",
}
`;

exports[`literal natural character set string with escapes 1`] = `
{
  "text": "n'hel\\'lo'",
  "type": "string",
}
`;

exports[`literal natural character set string with escapes 2`] = `
{
  "text": "N'hel''lo'",
  "type": "string",
}
`;

exports[`literal null 1`] = `
{
  "text": "null",
  "type": "null",
}
`;

exports[`literal null 2`] = `
{
  "text": "NULL",
  "type": "null",
}
`;

exports[`literal number 1`] = `
{
  "text": "0",
  "type": "number",
}
`;

exports[`literal number 2`] = `
{
  "text": "123",
  "type": "number",
}
`;

exports[`literal number 3`] = `
{
  "text": "0.15",
  "type": "number",
}
`;

exports[`literal number 4`] = `
{
  "text": "123e15",
  "type": "number",
}
`;

exports[`literal number 5`] = `
{
  "text": "1.23E+13",
  "type": "number",
}
`;

exports[`literal number 6`] = `
{
  "text": "0.107e-62",
  "type": "number",
}
`;

exports[`literal single-quoted string 1`] = `
{
  "text": "'hello'",
  "type": "string",
}
`;

exports[`literal single-quoted string with charset 1`] = `
{
  "charset": "binary",
  "string": {
    "text": "'hello'",
    "type": "string",
  },
  "type": "string_with_charset",
}
`;

exports[`literal single-quoted string with charset and comment 1`] = `
{
  "charset": "latin1",
  "string": {
    "leadingComments": [
      {
        "text": "/* comment */",
        "type": "block_comment",
      },
    ],
    "text": "'hello'",
    "type": "string",
  },
  "type": "string_with_charset",
}
`;

exports[`literal single-quoted string with escapes 1`] = `
{
  "text": "'hel\\'lo'",
  "type": "string",
}
`;

exports[`literal single-quoted string with escapes 2`] = `
{
  "text": "'hel''lo'",
  "type": "string",
}
`;
